mtags/src/main/scala/scala/meta/internal/mtags/SymbolIndexBucket.scala

```diff
@@ -0,0 +1,279 @@
+  // on the fly.
+  private def isTrivialToplevelSymbol(path: String, symbol: String): Boolean = {
+    val pathBuffer =
+      CharBuffer.wrap(path).subSequence(1, path.length - ".scala".length)
+    val symbolBuffer =
+      CharBuffer.wrap(symbol).subSequence(0, symbol.length - 1)
+    pathBuffer.equals(symbolBuffer)
+  }
+
+  def addToplevelSymbol(
+      path: String,
+      source: AbsolutePath,
+      toplevel: String
+  ): Unit = {
+    if (source.isAmmoniteScript || !isTrivialToplevelSymbol(path, toplevel)) {
+      val acc = toplevels.getOrElse(toplevel, Set.empty)
+      toplevels(toplevel) = acc + source
+    }
+  }
+
+  def query(symbol: Symbol): List[SymbolDefinition] =
+    query0(symbol, symbol)
+
+  /**
+   * Returns the file where symbol is defined, if any.
+   *
+   * Uses two strategies to recover from missing symbol definitions:
+   * - try to enter the toplevel symbol definition, then lookup symbol again.
+   * - if the symbol is synthetic, for examples from a case class of macro annotation,
+   *  fall back to related symbols from the enclosing class, see `DefinitionAlternatives`.
+   *
+   * @param querySymbol The original symbol that was queried by the user.
+   * @param symbol The symbol that
+   * @return
+   */
+  private def query0(
+      querySymbol: Symbol,
+      symbol: Symbol
+  ): List[SymbolDefinition] = {
+    if (!definitions.contains(symbol.value)) {
+      // Fallback 1: enter the toplevel symbol definition
+      val toplevel = symbol.toplevel
+      toplevels.get(toplevel.value) match {
+        case Some(files) =>
+          files.foreach(addMtagsSourceFile)
+        case _ =>
+          loadFromSourceJars(trivialPaths(toplevel))
+            .orElse(loadFromSourceJars(modulePaths(toplevel)))
+            .foreach(_.foreach(addMtagsSourceFile))
+      }
+    }
+    if (!definitions.contains(symbol.value)) {
+      // Fallback 2: guess related symbols from the enclosing class.
+      DefinitionAlternatives(symbol)
+        .flatMap(alternative => query0(querySymbol, alternative))
+    } else {
+      definitions
+        .get(symbol.value)
+        .map { paths =>
+          paths.map { p =>
+            SymbolDefinition(
+              querySymbol = querySymbol,
+              definitionSymbol = symbol,
+              path = p,
+              dialect = dialect
+            )
+          }.toList
+        }
+        .getOrElse(List.empty)
+    }
+  }
+  // similar as addSourceFile except indexes all global symbols instead of
+  // only non-trivial toplevel symbols.
+  private def addMtagsSourceFile(file: AbsolutePath): Unit = {
+    val docs: s.TextDocuments = PathIO.extension(file.toNIO) match {
+      case "scala" | "java" | "sc" =>
+        val language = file.toLanguage
+        val toIndexSource0 = toIndexSource(file).getOrElse(file)
+        val input = toIndexSource0.toInput
+        val document = mtags.index(language, input)
```

Comment:
We should also pass the dialect here to avoid the fallback in https://github.com/scalameta/metals/blob/12db95cb14b1fefdc982243f0e4a1a6311914304/mtags/src/main/scala/scala/meta/internal/mtags/ScalaMtags.scala#L27

Reply:
I will take a look at it